go/libkb/leveldb.go

```diff
@@ -11,56 +11,145 @@ import (
 
 	"github.com/syndtr/goleveldb/leveldb"
 	errors "github.com/syndtr/goleveldb/leveldb/errors"
+	"github.com/syndtr/goleveldb/leveldb/opt"
 )
 
+// table names
+const (
+	levelDbTableLo = "lo"
+	levelDbTableKv = "kv"
+)
+
+type levelDBOps interface {
+	Delete(key []byte, wo *opt.WriteOptions) error
+	Get(key []byte, ro *opt.ReadOptions) (value []byte, err error)
+	Put(key, value []byte, wo *opt.WriteOptions) error
+	Write(b *leveldb.Batch, wo *opt.WriteOptions) error
+}
+
+func levelDbPut(ops levelDBOps, id DbKey, aliases []DbKey, value []byte) error {
+	batch := new(leveldb.Batch)
+	idb := id.ToBytes(levelDbTableKv)
+	batch.Put(idb, value)
+	if aliases != nil {
+		for _, alias := range aliases {
+			batch.Put(alias.ToBytes(levelDbTableLo), idb)
+		}
+	}
+
+	return ops.Write(batch, nil)
+}
+
+func levelDbGetWhich(ops levelDBOps, id DbKey, which string) (val []byte, found bool, err error) {
+	val, err = ops.Get(id.ToBytes(which), nil)
+	found = false
+	if err == nil {
+		found = true
+	} else if err == leveldb.ErrNotFound {
+		err = nil
+	}
+	return val, found, err
+}
+
+func levelDbGet(ops levelDBOps, id DbKey) (val []byte, found bool, err error) {
+	return levelDbGetWhich(ops, id, levelDbTableKv)
+}
+
+func levelDbLookup(ops levelDBOps, id DbKey) (val []byte, found bool, err error) {
+	val, found, err = levelDbGetWhich(ops, id, levelDbTableLo)
+	if found {
+		if tab, id2, err2 := DbKeyParse(string(val)); err2 != nil {
+			err = err2
+		} else if tab != levelDbTableKv {
+			err = fmt.Errorf("bad alias; expected 'kv' but got '%s'", tab)
+		} else {
+			val, found, err = levelDbGetWhich(ops, *id2, levelDbTableKv)
+		}
+	}
+	return val, found, err
+}
+
+func levelDbDelete(ops levelDBOps, id DbKey) error {
+	return ops.Delete(id.ToBytes(levelDbTableKv), nil)
+}
+
 type LevelDb struct {
-	db       *leveldb.DB
+	// We use a RWMutes here to ensure close doesn't happen in the middle of
+	// other DB operations, and DB operations doesn't happen after close. The
+	// lock should be considered for the db pointer and dbOpenerOnce pointer,
+	// rather than the DB itself.  More specifically, close does Lock(), while
+	// other DB operations does RLock().
+	sync.RWMutex
```

Comment:
It's performance consideration; if we use a regular mutex, we'd have to do a regular lock on every operation, which essentially makes the DB single threaded